### ISSUE TITLE
[wishlist] ideal way to store "version"

### ISSUE THREAD

**Issue by yarikoptic** [singularityhub/singularityhub.github.io#47]
There is currently
```shell
yhalchen@discovery:/.singularity.d$ grep -i -r version .
./Singularity:OSVersion: stretch
./bootstrap/Singularity:OSVersion: stretch
./labels.json:    "SINGULARITY_DEFFILE_OSVERSION": "stretch",
./labels.json:    "SINGULARITY_BOOTSTRAP_VERSION": "2.3-master.gf737013"
```
so we know the version of the singularity was used but I don't think there is a way to determine within the container which version of the repository was used.  Ideally I would like to see a version which would be based on the "git describe --match=sing-*" output, so I could tag versions of the singularity definition, but then even if it changes without tagging, I would get a meaningful description from git describe:
```shell
$> git describe --tags --match sing-\* | sed -e 's,^sing-,,g'
2.2-1-g51e579c
```
so I could run that within %post, generate some file and add it somewhere under /.singularity.d but wondered first to ask so I do not reinvent the wheel etc ;)  @vsoch - ideas? ;)

**Comment by yarikoptic** [comment]
ideally it should get integrated somehow with version in the image filename:
```
$> singularity pull shub://neurodebian/neurodebian
Progress |===================================| 100.0% 
Done. Container is at: ./neurodebian-neurodebian-master.img
singularity pull shub://neurodebian/neurodebian  82.80s user 19.72s system 84% cpu 2:01.07 total
```
so that it could be  neurodebian-neurodebian-2.2-1-g51e579c.img  instead? my point is that image is re-generated, users wouldn't know which one they are actually using -- not good for reproducibility

**Comment by vsoch** [comment]
@yarikoptic is it not sufficient to have it represented as a label? Eg: `2.3-master.gf737013` --> `[singularity-version][branch][commit start]`

**Comment by yarikoptic** [comment]
it is not sortable, it is not understood by git

sortable: Would be impossible to figure out which one is "later".
Sure thing, because of the possible branches, comparison by # of commits since last one would not necessarily be correct across branches, but since you are specifying the branch -- it makes it nice and dandy.  So ideally, it should have been 2.3-2-master-gf7370131

understood by git:  why above form -- so that "git" undestands it as well:

```shell
$> git show 2.3-2-master-gf7370131 | head -n 2
commit f737013143f9a5e54ca01a0ba398722909c3c37b
Merge: adf5259a 64aa238c
```
whenever current one is not undestood by git:
```shell
$> git show 2.3-master.gf737013 | head -n 2
fatal: ambiguous argument '2.3-master.gf737013': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
```

you can actually see that specifying a branch somewhat sticks it out (note that branch is just a name, which could be arbitrary across clones, and thus not uniquely identifying it)

**Comment by vsoch** [comment]
why don't you just pull with the name by commit command?  try pull --commit and see pull --help for details.

**Comment by yarikoptic** [comment]
those options are indeed nice to have, BUT require user to know the commit.  What I am aiming at is for naive user to not get mixed up -- do the pull, run some analysis using one version, then do pull again, run another analysis using another version without realizing that now he is using another (which ?) version of the image.  Having an explicit version (may be based on another cmdline switch) would be beneficial.  Also if that version is stored somewhere in env var, would be great for any tool which tracks provenance of commands execution (e.g. reprozip, etc) to capture that information, so later it would become easier to see how two environments differed. 
Original use case though was to print out the version in the  `%runscript` where for now I am doing it manually/hardcoded way:

```shell
%runscript
    echo "Welcome to the NeuroDebian v 2.2 (Debian stretch) environment"
```
whenever I want something like

```shell
%runscript
    echo "Welcome to the NeuroDebian v $SINGULARITY_IMAGE_VERSION (Debian stretch) environment"
```
to make it clear for the user which version he/she uses

makes sense?

**Comment by vsoch** [comment]
The user doesn't need to know the commit! For example, here is the standard way of pulling:

```
singularity pull shub://vsoch/hello-world
Done. Container is at: /home/vanessa/Documents/Dropbox/Code/shub/sregistry/vsoch-hello-world-master.img
```
Now let me grab the image but name it by the FULL commit, without needing to know in advance:

```
singularity pull --commit shub://vsoch/hello-world
Done. Container is at: /home/vanessa/Documents/Dropbox/Code/shub/sregistry/e279432e6d3962777bb7b5e8d54f30f4347d867e.img
```
What Singularity Hub can do is provide an easy way to ask for containers that pertain to a specific version. If you want customizability of this you can use the API (or the wrapper in Singularity Python) directly. To be clear, a version for Singularity Hub is a Github commit. What you are showing is the neurodebian version, which arguably you should store as a container label, and then you can parse the label from the metadata folder at runtime to get the version (see `/.singularity.d/labels.json`) 

Now to think about responsibility, or limits of what each component can do. What Singularity can do is keep track of it's version, and any labels that you add to it. We do that with the --version, which has the version of Singularity and the exact start of the commit, along with being able to specify `%labels`. Those two things alone should let you specify versioning in your container, at least with regards to Singularity and your software.  But Singularity doesn't know about Github, that's where Singularity Hub versioning comes in.  Singularity Hub will make it possible to pull a container and get its commit, or use the API directly to do the same. What Singularity Hub will not do is to stick it's fingers into your container and add other labels, or some kind of label specific to neurodebian, for example. Arguably, we could add some builder option to add labels, but I would argue that need is better met using the `%labels` section.

I think what you are asking is for Singularity Hub to write a custom label or environment variable into the container? This would mean editing the content of the container on the fly (as a builder) so if someone were to build the Singularity file without Singularity Hub, it wouldn't be the same. I don't like that.

Why not write a command in the definition recipe to, if it finds a github repo, to get the current commit and save it to an environment variable for the container? Or pull the container and name by commit? It must be the case that if someone builds the definition recipe at that particular commit, the same container is generated. Adding custom additions via the Singularity Hub builder does not ensure that.

**Comment by yarikoptic** [comment]
> --commit

I misunderstood its meaning -- I thought it was to request an image for that commit. Cool.

> %labels

for that I guess I should generate a singularity definition file with that version.  I feel that we want to avoid that and have it done within %post ?

> Why not write a command in the definition recipe to, if it finds a github repo, to get the current commit and save it to an environment variable for the container? 

that is the question -- how to do that "the best way" and to which variable?  I guess I could always just dump it to some file and then source it within %runscript.  But I thought there could be a better way.

**Comment by vsoch** [comment]
@yarikoptic I think for %labels, you would easily be able to add the label - by the time we are in post, the Docker bits (and the file for the labels at `/.singularity.d/labels.json' has been written. It's just a json file, so if you have a small script to read it in and add a label, you are then done! You could also just write a line to export the variable to the environment, we just added that in the development branch.

For your last question, it's up to you. Do you want a label (metadata), an environment variable, or both? I'd say both, because they have different use cases.

**Comment by yarikoptic** [comment]
greedy me says "both" too ;) ok -- will add it to labels.json for now, and use that information within %runscript.

**Comment by vsoch** [comment]
sounds good!

**Comment by yarikoptic** [comment]
ok, tried:  log is https://singularity-hub.org/containers/2517/log
```shell
+ sed -e s,^sing-,,g
+ git describe --tags --match sing-*
/bin/sh: 1: git: not found
+ v=
+ python -c import json; f='/.singularity.d/labels.json'; j=json.load(open(f)); j['SINGULARITY_IMAGE_VERSION']='' or '0.0.unknown';json.dump(j, open('/tmp/j','w'),indent=2)
Traceback (most recent call last):
File "", line 1, in
IOError: [Errno 2] No such file or directory: '/.singularity.d/labels.json'
ABORT: Aborting with RETVAL=255
```

and what is interesting I ended up without any  '/.singularity.d/labels.json' at the end (not even the ones singularity itself should place)

1. ideally if one command fails, the script/build should fail.  I guess I will try adding 'set -eu' now to %post to see if it has desired effect
2. when /.singularity.d is created?  even later than `%post`.  I could mkdir but then I guess that file would be overriden with new labels etc

**Comment by vsoch** [comment]
Ah, if you are using debootstrap (and not docker) then the labels aren't written until after `%post`, in which case you should just echo to the file (and create it). If you do a docker bootstrap, the file is likely to already be there, at least if the container has labels.

TLDR: don't try to add to the file, try to write to it! Let's try that? And I'm not sure this is a singularityhub issue, but I'd probably help you on the other board anyway, lol.

**Comment by yarikoptic** [comment]
ok -- my scripts worked but assumption was wrong -- sure thing within that %post script I am not within git repo , so could not invoke `git describe`...   is there any point in which code is executed while under original git repo?

**Comment by vsoch** [comment]
Yes, in the %setup section you can access your host, and the image base is at `$SINGULARITY_ROOTFS`

**Comment by yarikoptic** [comment]
woohoo
```shell
$> ./99229f67c6fa7ef144f52807473c2675a4a37751.img 
Welcome to the NeuroDebian v 2.2-7-g99229f6 (Debian stretch) environment
```
will add it now to the proper one ;)
